cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -1104,23 +1118,101 @@ private ArtifactCell(Id.Artifact artifactId) {
   private static class PluginData {
     private final PluginClass pluginClass;
     private final ArtifactRange usableBy;
-    private final URI artifactLocationURI;
+    private final String artifactLocationPath;
 
     PluginData(PluginClass pluginClass, ArtifactRange usableBy, Location artifactLocation) {
       this.pluginClass = pluginClass;
       this.usableBy = usableBy;
-      this.artifactLocationURI = artifactLocation.toURI();
+      this.artifactLocationPath = artifactLocation.toURI().getPath();
+    }
+
+    PluginData(PluginClass pluginClass, ArtifactRange usableBy, String artifactLocationPath) {
+      this.pluginClass = pluginClass;
+      this.usableBy = usableBy;
+      this.artifactLocationPath = artifactLocationPath;
     }
   }
 
   // Data that will be stored for an application class.
   private static class AppData {
     private final ApplicationClass appClass;
-    private final URI artifactLocationURI;
+    private final String artifactLocationPath;
 
     AppData(ApplicationClass appClass, Location artifactLocation) {
       this.appClass = appClass;
-      this.artifactLocationURI = artifactLocation.toURI();
+      this.artifactLocationPath = artifactLocation.toURI().getPath();
+    }
+
+    AppData(ApplicationClass appClass, String artifactLocationPath) {
+      this.appClass = appClass;
+      this.artifactLocationPath = artifactLocationPath;
+    }
+  }
+
+  private static class ArtifactDataDeserializer implements JsonDeserializer<ArtifactData> {
+    @Override
+    public ArtifactData deserialize(JsonElement json,
+                                    Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+      JsonObject jsonObj = json.getAsJsonObject();
+      String locationPath = null;
+      // For Backward Compatibility
+      if (jsonObj.has("locationURI")) {
+        try {
+          URI locationURI = new URI(jsonObj.get("locationURI").getAsString());
+          locationPath = locationURI.getPath();
+        } catch (Exception e) {
```

Comment:
You should only catch what can be thrown. In this case, catch URISyntaxException. Though this may be a moot point if you get the URI using the context.